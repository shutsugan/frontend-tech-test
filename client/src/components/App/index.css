:root {
  --main: #000;
  --secondary: #f1696a;
  --border: #ddd;
  --title: #001;
  --sub-title: #666;
  --label: #aaa;
  --radius: 3px;
  --transition: cubic-bezier(.25,.8,.25,1);
  --duration: .15s;
}

* {
  box-sizing: border-box;
}

html, body {
  margin: 0;
  padding: 0;
  font-family: Rajdhani, sans-serif;
}

a {
  text-decoration: none;
}

.flex {
  display: flex;
}

.flex-column {
  flex-direction: column;
}

.center {
  justify-content: center;
  align-items: center;
}

.start {
  justify-content: flex-start;
  align-items: flex-start;
}

.end {
  justify-content: flex-end;
  align-items: flex-end;
}

.space-between {
  justify-content: space-between;
  align-items: center;
}

.stretch {
  align-items: stretch;
}

.full {
  width: 100%;
}

.half {
  width: 60%;
}

.ht-80 {
  height: 80px;
}

.pd-16 {
  padding: 16px;
}

.pdt-16 {
  padding-top: 16px;
}

.pd-8 {
  padding: 8px;
}

.pdt-8 {
  padding: 8px 0;
}

.mr-none {
  margin: 0;
}

.mrb-none {
  margin-bottom: 0;
}

.mrb-16 {
  margin-bottom: 16px;
}

.mrt-16 {
  margin-top: 16px;
}

.mrl-16 {
  margin-left: 16px;
}

.mrr-16 {
  margin-right: 16px;
}

.mrs-8 {
  margin: 0 8px;
}

.mr-auto {
  margin: auto;
}

.button {
  text-transform: uppercase;
  text-align: center;
  font-size: .8rem;
  font-weight: 600;
  background: var(--main);
  color: #fff;
  border: none;
  border-radius: var(--radius);
  cursor: pointer;
  transition: box-shadow var(--duration) var(--transition);
}

.button:hover {
  box-shadow:
    0 1px 3px rgba(0,0,0,0.12),
    0 1px 2px rgba(0,0,0,0.24);
}

.banner-button {
  background: transparent;
  border: solid 2px var(--main);
  color: #000;
  font-size: .7rem;
  text-decoration: none;
}

.link {
  text-decoration: none;
  color: var(--secondary);
  transition: color var(--duration) var(--transition);
  cursor: pointer;
}

.link:hover {
  color: var(--main);
}

.relative {
  position: relative;
}

.absolute {
  position: absolute;
}

.card {
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
}

.br-bottom {
  border-bottom: dashed .5px var(--border);
}

.br-bottom:last-child {
  border-bottom: none;
}

.fade-appear {
  opacity: 0.01;
}

.fade-appear.fade-appear-active {
  opacity: 1;
  transition: opacity var(--duration) var(--transition);
}

.fade-enter {
  opacity: 0.01;
}

.fade-enter.fade-enter-active {
  opacity: 1;
  transition: opacity var(--duration) var(--transition);
}

.fade-leave {
  opacity: 1;
}

.fade-leave.fade-leave-active {
  opacity: 0.01;
  transition: opacity var(--duration) var(--transition);
}

@media only screen and (max-width: 615px) {
  .half {
    width: 90%;
  }
}