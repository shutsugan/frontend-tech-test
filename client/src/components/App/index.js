import React from 'react';

import './index.css';

const App = _ => {

  return (
    <div className="app">the app</div>
  );
};

export default App;
