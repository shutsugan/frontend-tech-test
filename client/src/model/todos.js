import { action, computed, thunk } from "easy-peasy";

const base_url = 'http://localhost:9001';

const noop = _ => {};
export const save = async (axios, actions, method, url, callback = noop) => {
    const res = await axios[method](`${base_url}/task/${url}`);
    const { task, message } = await res.data;

    callback(task);
    actions.setMessage(message);
};

export const checkProps = (actions, title, description) => {
    if (!title && !description) return actions.setMessage('Fields are reuqired');
};

export default {
    todos: [],
    selectedTodo: null,
    message: null,

    list: computed(state => Object.values(state.todos)),

    //actions
    setMessage: action((state, message) => state.message = message),
    selectTodo: action((state, todo) => state.selectedTodo = todo),
    added: action((state, todo) => state.todos[todo.id] = todo),

    fetched: action((state, payload = []) => {
        state.todos = payload.reduce((acc, todo) => {
            acc[todo.id] = todo;

            return acc;
        }, {});
    }),

    //thunks
    fetchTodos: thunk(async (actions, _, { injections }) => {
        const { axios } = injections
        const res = await axios.get(`${base_url}/tasks`);
        const { tasks } = await res.data;

        actions.fetched(tasks);
    }),

    addTodo: thunk(async (actions, { title, description }, { injections }) => {
        const { axios } = injections

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'post',
            `create/${title}/${description}`,
            todo => actions.added(todo)
        );
    }),

    updateTodo: thunk(async (actions, { id, title, description }, { injections }) => {
        const { axios } = injections;

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'put',
            `update/${id}/${title}/${description}`,
            _ => {
                actions.fetchTodos();
                actions.selectTodo(null);
            }
        );
    }),

    removeTodo: thunk(async (actions, id, { injections }) => {
        const { axios } = injections;

        save(
            axios,
            actions,
            'delete',
            `delete/${id}`,
            _ => actions.fetchTodos()
        );
    })
};