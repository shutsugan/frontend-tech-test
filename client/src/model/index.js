import todos from './todos';
import { thunk } from 'easy-peasy';

export default {
    todos,
    initialiseTodos: thunk(async (actions, payload, { dispatch }) => {
        await dispatch.todos.fetchTodos();
    })
};