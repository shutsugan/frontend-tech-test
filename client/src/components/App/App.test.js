import React from 'react';
import renderer from 'react-test-renderer';
import { createStore, StoreProvider } from 'easy-peasy';
import { expect as chaiExpect } from 'chai';

import model from '../../model';
import App from './index';

const store = createStore(
    model, { injections: { axios: jest.fn() } }
);

describe('App component', () => {
    it ('should render the app component', () => {
        const tree = renderer.create(
            <StoreProvider store={store}>
                <App />
            </StoreProvider>
        ).toJSON();
        
        expect(tree).toMatchSnapshot();
        chaiExpect(tree).to.be.an('object');
        chaiExpect(tree).to.have.own.property('props');
        chaiExpect(tree.children.length).to.equal(2);
    });
})