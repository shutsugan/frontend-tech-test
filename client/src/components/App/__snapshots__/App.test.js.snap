// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render the app component 1`] = `
<div
  className="app half mr-auto flex-column center pd-16"
>
  <div
    className="todo-list card full flex-column mrb-16"
  />
  <div
    className="add-todo full"
  >
    <div
      className="fleid flex-column start mrb-16 relative"
    >
      <label
        className="required"
      >
        Title
      </label>
      <input
        aria-label="title"
        className="field__input full pd-16"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="fleid flex-column start mrb-16 relative"
    >
      <label
        className="required"
      >
        Description
      </label>
      <textarea
        aria-label="description"
        className="field__input full pd-16 ht-80"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      aria-label="button"
      className="button full pd-16"
      onClick={[Function]}
    >
      Add Todo
    </button>
  </div>
</div>
`;
